gateway/handler_unixfs__redirects.go

```diff
@@ -214,55 +230,53 @@ func isUnixfsResponseFormat(responseFormat string) bool {
 // Deprecated: legacy ipfs-404.html files are superseded by _redirects file
 // This is provided only for backward-compatibility, until websites migrate
 // to 404s managed via _redirects file (https://github.com/ipfs/specs/pull/290)
-func (i *handler) serveLegacy404IfPresent(w http.ResponseWriter, r *http.Request, contentPath ipath.Path) bool {
-	resolved404Path, ctype, err := i.searchUpTreeFor404(r, contentPath)
-	if err != nil {
-		return false
-	}
-
-	dr, err := i.api.GetUnixFsNode(r.Context(), resolved404Path)
+func (i *handler) serveLegacy404IfPresent(w http.ResponseWriter, r *http.Request, imPath ImmutablePath) bool {
+	resolved404File, ctype, err := i.searchUpTreeFor404(r, imPath)
 	if err != nil {
 		return false
 	}
-	defer dr.Close()
-
-	f, ok := dr.(files.File)
-	if !ok {
-		return false
-	}
+	defer resolved404File.Close()
 
-	size, err := f.Size()
+	size, err := resolved404File.Size()
 	if err != nil {
 		return false
 	}
 
-	log.Debugw("using pretty 404 file", "path", contentPath)
+	log.Debugw("using pretty 404 file", "path", imPath)
 	w.Header().Set("Content-Type", ctype)
 	w.Header().Set("Content-Length", strconv.FormatInt(size, 10))
 	w.WriteHeader(http.StatusNotFound)
-	_, err = io.CopyN(w, f, size)
+	_, err = io.CopyN(w, resolved404File, size)
 	return err == nil
 }
 
-func (i *handler) searchUpTreeFor404(r *http.Request, contentPath ipath.Path) (ipath.Resolved, string, error) {
+func (i *handler) searchUpTreeFor404(r *http.Request, imPath ImmutablePath) (files.File, string, error) {
 	filename404, ctype, err := preferred404Filename(r.Header.Values("Accept"))
 	if err != nil {
 		return nil, "", err
 	}
 
-	pathComponents := strings.Split(contentPath.String(), "/")
+	pathComponents := strings.Split(imPath.String(), "/")
 
 	for idx := len(pathComponents); idx >= 3; idx-- {
 		pretty404 := gopath.Join(append(pathComponents[0:idx], filename404)...)
 		parsed404Path := ipath.New("/" + pretty404)
 		if parsed404Path.IsValid() != nil {
 			break
 		}
-		resolvedPath, err := i.api.ResolvePath(r.Context(), parsed404Path)
+		imparsed404Path, err := NewImmutablePath(parsed404Path)
+		if err != nil {
+			break // TODO: followed the same pattern with IsValid, should these be logging?
```

Comment:
ok since ipfs-404.html already did this, and is deprecated undocumented behavior anyway.

```suggestion
			break
```